src/ScottPlot5/ScottPlot5/DataSourceUtilities.cs

```diff
@@ -0,0 +1,475 @@
+﻿using ScottPlot.Collections;
+using ScottPlot.DataSources;
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Runtime.CompilerServices;
+
+namespace ScottPlot
+{
+    public sealed class BinarySearchComparer : IComparer<Coordinates>, IComparer<double>, IComparer<RootedCoordinateVector>, IComparer<RootedPixelVector>
+    {
+        public static readonly BinarySearchComparer Instance = new BinarySearchComparer();
+
+        public static IComparer<T> GetComparer<T>() => GenericComparer<T>.Instance;
+        public int Compare(Coordinates a, Coordinates b) => a.X.CompareTo(b.X);
+        public int Compare(double a, double b) => a.CompareTo(b);
+        public int Compare(RootedPixelVector x, RootedPixelVector y) => x.Point.X.CompareTo(y.Point.X);
+        public int Compare(RootedCoordinateVector x, RootedCoordinateVector y) => x.Point.X.CompareTo(y.Point.X);
+
+    }
+
+    public sealed class GenericComparer<T> : IComparer<T>
+    {
+        ///<summary>
+        ///For standard numerics (single, long, int, etc) this will be '<see cref="Comparer{T}.Default"/>', which is significantly faster than this interface implementation
+        ///</summary> 
+        public static readonly IComparer<T> Instance = GetComparer(); 
+
+#pragma warning disable CS8767 // handled by null check
+        public int Compare(T a, T b)
+        {
+            return true switch
```

Comment:
This was done for readability for me mostly.

It was submitted checking both variables because compile warnings prevented the GitHub build (despite compiling fine). The secondary check (if b is coordinate) should be JIT'd out because if A is true then B is always true. 

All that said, I'll throw this into a benchmark and run some tests and update with best result.

Reply:
Leaving as is! 

```
private GenericComparer<double> Comparer = new GenericComparer<double>();

[Benchmark]
public void CompareSwitch() => Comparer.Compare(0, 1); // as is

[Benchmark]
public void CompareIf() => Comparer.Compare2(0, 1); // re-written using IF statements
```

Benchmark results:
| Method        | Mean     | Error     | StdDev    | Median   | Gen0   | Allocated |
|-------------- |---------:|----------:|----------:|---------:|-------:|----------:|
| CompareSwitch | 1.335 ns | 0.0552 ns | 0.1312 ns | 1.318 ns |      - |         - |
| CompareIf     | 3.506 ns | 0.0936 ns | 0.2349 ns | 3.419 ns | 0.0038 |      24 B |

Reply:
Removed this completely, I realized GenericComparer does not need to implement the interface, it just needs to provide one.